engine/access/rpc/backend/scripts/scripts.go

```diff
@@ -72,67 +112,232 @@ func NewScriptsBackend(
 	script []byte,
 	arguments [][]byte,
-) ([]byte, error) {
+	userCriteria optimistic_sync.Criteria,
+) ([]byte, *accessmodel.ExecutorMetadata, error) {
 	if !commonrpc.CheckScriptSize(script, arguments, b.maxScriptAndArgumentSize) {
-		return nil, status.Error(codes.InvalidArgument, commonrpc.ErrScriptTooLarge.Error())
+		return nil, nil, access.NewInvalidRequestError(commonrpc.ErrScriptTooLarge)
 	}
 
 	header, err := b.headers.ByHeight(blockHeight)
 	if err != nil {
-		return nil, commonrpc.ConvertStorageError(common.ResolveHeightError(b.state.Params(), blockHeight, err))
+		err = access.RequireErrorIs(ctx, err, storage.ErrNotFound)
+		err = common.ResolveHeightError(b.state.Params(), blockHeight, err)
+		err = fmt.Errorf("failed to find header by height: %w", err)
+		return nil, nil, access.NewDataNotFoundError("header", err)
+	}
+
+	headerID := header.ID()
+	executionResultInfo, err := b.executionResultProvider.ExecutionResultInfo(headerID, userCriteria)
+	if err != nil {
+		err = fmt.Errorf("failed to get execution result info for block: %w", err)
+		switch {
+		case errors.Is(err, storage.ErrNotFound):
+			return nil, nil, access.NewDataNotFoundError("execution data", err)
+		case common.IsInsufficientExecutionReceipts(err):
+			return nil, nil, access.NewDataNotFoundError("execution data", err)
+		default:
+			return nil, nil, access.RequireNoError(ctx, err)
+		}
 	}
 
-	res, _, err := b.executor.Execute(ctx, executor.NewScriptExecutionRequest(header.ID(), blockHeight, script, arguments))
-	return res, err
+	res, metadata, err := b.executor.Execute(
+		ctx,
+		executor.NewScriptExecutionRequest(headerID, blockHeight, script, arguments),
+		executionResultInfo,
+	)
+	if err != nil {
+		return nil, nil, handleScriptExecutionError(ctx, err)
+	}
+
+	return res, metadata, nil
+}
+
+// handleScriptExecutionError converts storage, execution, version or gRPC errors
+// into access-layer errors according to the Access API error handling convention.
+//
+// Expected error returns during normal operation:
+//   - [access.InvalidRequestError] - if the script execution failed due to invalid arguments or runtime errors.
+//   - [access.ResourceExhausted] - if computation or memory limits were exceeded.
+//   - [access.DataNotFoundError] - if the script data or related execution data was not found,
+//     or if the block height is out of range or incompatible with node version.
+//   - [access.OutOfRangeError] - if the requested data is outside the available range.
+//   - [access.PreconditionFailedError] - if data for block is not available.
+//   - [access.RequestCanceledError] - if the script execution was canceled.
+//   - [access.RequestTimedOutError] - if the script execution timed out.
+//   - [access.ServiceUnavailable] - if no nodes are available or a connection could not be established.
+//   - [access.InternalError] - for internal failures or conversion errors.
+func handleScriptExecutionError(ctx context.Context, err error) error {
```

Comment:
I don't think we should replicate all of the errors from access for each endpoint. I said before that I think we should use backend specific errors and do a mapping within the backend logic to the correct access error. If we're going to have to create a ton of single use sentinel errors, it probably makes sense to just have the `Execute` method do the conversion to the access errors. then we could just assert `RequireAccessError()` in the backend endpoint to ensure it was done by the provider.